src/tyr/isochrone_serializer.cc

```diff
@@ -12,40 +12,183 @@ using namespace valhalla::baldr::json;
 
 namespace {
 using rgba_t = std::tuple<float, float, float>;
+
+using namespace valhalla;
+using namespace tyr;
+using namespace midgard;
+using contour_t = std::list<PointLL>;                 // single ring
+using feature_t = std::list<contour_t>;               // rings per interval
+using contours_t = std::vector<std::list<feature_t>>; // all rings
+using contour_group_t = std::vector<const contour_t*>;
+using grouped_contours_t = std::vector<contour_group_t>;
+using contour_interval_t = std::tuple<size_t, float, std::string, std::string>;
+
+grouped_contours_t GroupContours(const bool polygons, const feature_t& contours) {
+  grouped_contours_t results;
+
+  // if the user requested linestrings, we'll give them linestrings
+  if (!polygons || contours.size() < 2) {
+    std::for_each(contours.begin(), contours.end(),
+                  [&results](const contour_t& c) { results.push_back({&c}); });
+    return results;
+  }
+
+  // inner rings have negative area
+  auto isExteriorRing = [](const contour_t& c) -> bool { return polygon_area(c) > 0; };
+
+  std::vector<const contour_t*> inner_ptrs;
+  std::for_each(contours.begin(), contours.end(),
+                [&results, &inner_ptrs, isExteriorRing](const contour_t& c) {
+                  if (isExteriorRing(c)) {
+                    results.push_back({&c});
+                  } else {
+                    inner_ptrs.push_back(&c);
+                  }
+                });
+
+  // exactly one exterior ring, so all inners go in one group
+  if (results.size() == 1) {
+    std::for_each(inner_ptrs.begin(), inner_ptrs.end(),
+                  [&inner_ptrs, &results](const contour_t* c) { results[0].push_back(c); });
+    return results;
+  }
+
+  // now we're dealing with multiple exterior rings
+  // iterate over outer rings and for each inner ring check if the inner ring is within the outer ring
+  for (const auto inner : inner_ptrs) {
+    // construct bbox
+    AABB2<PointLL> inner_bbox(std::vector(inner->rbegin(), inner->rend()));
+    bool found_exterior;
+    // go over exterior rings from smallest to largest
+    for (size_t i = results.size(); i > 0; --i) {
+      AABB2<PointLL> outer_bbox(std::vector(results[i - 1][0]->rbegin(), results[i - 1][0]->rend()));
+
+      // contain check
+      if (outer_bbox.Contains(inner_bbox)) {
```

Comment:
this is the easy way, but would in some cases lead to "misgrouped" inner rings. the better solution would be to do actual polygon-within-polygon checks

Reply:
use point in polygon